core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -130,63 +126,104 @@ inline void check_shmem_request(CudaInternal const* cuda_instance, int shmem) {
+
+  // Compute how many threads per sm we actually want
+  const size_t maxthreads_per_sm = device_props.maxThreadsPerMultiProcessor;
+  // only allocate multiples of warp size
+  const size_t numthreads_desired =
+      ((maxthreads_per_sm * occupancy / 100 + 31) / 32) * 32;
+  // Get close to the desired occupancy,
+  // don't undershoot by much but also don't allocate a whole new block just
+  // because one is a few threads over otherwise.
+  size_t numblocks_desired =
+      (numthreads_desired + block_size * 0.8) / block_size;
+  numblocks_desired = ::std::min(max_blocks_regs, numblocks_desired);
+  if (numblocks_desired == 0) numblocks_desired = 1;
+
+  // Calculate how much shared memory we need per block
+  size_t shmem_per_block = shmem + func_attr.sharedSizeBytes;
+  // The minimum shared memory allocation we can have in total per SM is 8kB
+  // if we want to lower occupancy we have to make sure we at least request that
+  // much in aggregate over all blocks, so that shared memory actually becomes a
+  // limiting factor for occupancy
+  constexpr size_t min_shmem_size_per_sm = 8192;
+  if ((occupancy < 100) &&
+      (shmem_per_block * numblocks_desired < min_shmem_size_per_sm)) {
+    shmem_per_block = min_shmem_size_per_sm / numblocks_desired;
+    shmem           = shmem_per_block - func_attr.sharedSizeBytes;
+  }
+
+  // Compute the carveout fraction we need based on occupancy
+  // Use multiples of 8kB
+  const size_t maxshmem_per_sm = device_props.sharedMemPerMultiprocessor;
+  size_t carveout              = shmem_per_block == 0
+                        ? 0
+                        : 100 *
+                              (((numblocks_desired * shmem_per_block +
+                                 min_shmem_size_per_sm - 1) /
+                                min_shmem_size_per_sm) *
+                               min_shmem_size_per_sm) /
+                              maxshmem_per_sm;
+  if (carveout > 100) carveout = 100;
+
+  // Set the carveout, but only call it once per kernel or when it changes
   auto set_cache_config = [&] {
-    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFuncSetCacheConfig(
-        func,
-        (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)));
-    return prefer_shmem;
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFuncSetAttribute(
+        func, cudaFuncAttributePreferredSharedMemoryCarveout, carveout));
+    return carveout;
   };
-  static bool cache_config_preference_cached = set_cache_config();
-  if (cache_config_preference_cached != prefer_shmem) {
+  // Store the value in a static variable so we only reset if needed
+  static int cache_config_preference_cached = set_cache_config();
+  if (cache_config_preference_cached != carveout) {
     cache_config_preference_cached = set_cache_config();
   }
 #else
   // Use the parameters so we don't get a warning
   (void)func;
-  (void)prefer_shmem;
+  (void)device_props;
+  (void)block_size;
+  (void)occupancy;
```

Comment:
We've got C++17 now - these variables can just be `[[maybe_unused]]`

Or are you thinking that even this might get picked over the a 3.7 release?

Reply:
maybe though unlikely. But that was the scheme from before, I didn't actively try to modernize for C++17 in the same go.